### DISCUSSION TITLE
Content localization within decoupled website

### DISCUSSION THREAD

**Discussion by draxtor** [OrchardCMS/OrchardCore#7649] [Q&A]
Hello,
I've created decoupled website using Asp.Net Core 3.1 
I'm using orchard core nugets:
![image](https://user-images.githubusercontent.com/3110724/99191743-f4178500-276e-11eb-9b90-8dab654cf959.png)
2 abstraction nugets I've installed out of despair, I don't know what they do I was looking for ILocalizationService which is mentioned on the internet somewhere.
 
I turned on all localization features and created little component for switching languages (eng and fra for now). 
Based on 
`<html lang="@Orchard.CultureName()">`
 it looks that it is working fine.
So I started to localize content types. But I cannot found any way of filtering content items based on culture. 
I've tried:
 ```
   var articles = await _orchard.GetRecentContentItemsByContentTypeAsync("HomeArticles");
   var articles = await _orchard.QueryContentItemsAsync(x=> x.Where(y => y.Published));
```
but all items for all languages are returned.
There are no overload which I can use to send culture info.

Is there a way to use content localization in decoupled website? Or at least to filet content items based on culture.

Thanks

**Comment by hishamco** [comment]
Try to install Content Localization module if you interested in localizable content items

**Reply by draxtor** [reply]
As I said in my question I already did install Content Localization module and:

- I've read: https://docs.orchardcore.net/en/dev/docs/reference/modules/ContentLocalization/ and enabled it
![image](https://user-images.githubusercontent.com/3110724/99194160-14027500-277e-11eb-8b19-7708ad50a03c.png)
- I've added 2 cultures in the settings  
![image](https://user-images.githubusercontent.com/3110724/99194150-fa612d80-277d-11eb-8e2e-a9b316953c74.png)
- I've set up a content for one of my content types  in different languages.
![image](https://user-images.githubusercontent.com/3110724/99194173-25e41800-277e-11eb-9f92-fc0669595632.png)

But I cannot do anything with it except loading all of them iterating through each of them and check LocalizationPart. 
I'm using IOrchardHelper for fetching content and it does not differentiates different cultures.

**Reply by hishamco** [reply]
Checking `LocalizationPart` should work but you could use `IOrchardHelper.GetContentCultureAsync()` instead, indeed we need a helper method to filter them by culture, @jptissot do we have something achieve this?

**Reply by hishamco** [reply]
After a quick look seems `DefaultContentLocalizationManager` has some helper method that allow you to achieve that

**Reply by draxtor** [reply]
Cool thanks I'll try it.
I have some progress with SQL queries but it is cumbersome to glue the whole object together with SQL query. Alternative is using SQL to get id and then do the helper stuff (with n+1 penalty for fetching).

**Reply by draxtor** [reply]
I've tried DefaultContentLocalizationManager but it doesn't fit into my scenario.
It exposes:
 ```
       Task<IDictionary<string, ContentItem>> DeduplicateContentItemsAsync(IEnumerable<ContentItem> contentItems);
        Task<ContentItem> GetContentItemAsync(string localizationSet, string culture);
        Task<IDictionary<string, string>> GetFirstItemIdForSetsAsync(IEnumerable<string> localizationSets);
        Task<IEnumerable<ContentItem>> GetItemsForSetAsync(string localizationSet);
        Task<IEnumerable<ContentItem>> GetItemsForSetsAsync(IEnumerable<string> localizationSets, string culture);
        Task<ContentItem> LocalizeAsync(ContentItem content, string targetCulture);
```

Problem is that localizationSet is required for all of them except for the DeduplicateContentItemsAsync one. I don't have localizationSet at the moment I have ContentType and culture.

On the other hand SQL Query is fine, not sure about performance but I created following query in the admin:
```
SELECT * FROM 
[ContentItemIndex] 
INNER JOIN [LocalizedContentItemIndex] ON [ContentItemIndex].[ContentItemId] = [LocalizedContentItemIndex].[ContentItemId]
INNER JOIN [Document] ON [ContentItemIndex].[DocumentId] = [Document].[Id] 
WHERE [ContentItemIndex].[ContentType] = @contentType AND [LocalizedContentItemIndex].[Culture] = @lang;

```
and it's fulfilling my use case.

**Reply by jptissot** [reply]
Have a look at what is done in https://github.com/OrchardCMS/OrchardCore/blob/6cddd9f9904ff117ba82587f39fc76f49fac3cc3/src%2FOrchardCore.Modules%2FOrchardCore.ContentLocalization%2FLiquid%2FContentLocalizationFilter.cs

**Reply by ManOguaR** [reply]
.With<LocalizedContentItemIndex>(x => x.Culture == culture)
